matrix/Dual.hpp

```diff
@@ -0,0 +1,361 @@
+/**
+ * @file Dual.hpp
+ *
+ * This is a dual number type for calculating automatic derivatives.
+ *
+ * Based roughly on the methods described in:
+ * Automatic Differentiation, C++ Templates and Photogrammetry, by Dan Piponi
+ * and
+ * Ceres Solver's excellent Jet.h
+ *
+ * @author Julian Kent <julian@auterion.com>
+ */
+
+#pragma once
+
+#include "math.hpp"
+
+namespace matrix
+{
+
+template <typename Type, size_t M>
+class Vector;
+
+template <typename Scalar, size_t N>
+struct Dual
+{
+    static constexpr size_t WIDTH = N;
+
+    Scalar value {};
+    Vector<Scalar, N> derivative;
+
+    Dual() = default;
+
+    explicit Dual(Scalar v, size_t inputDimension = 65535)
```

Comment:
It is not fully clear to me what is happening in this constructor. Can you give an example?

Reply:
Good catch, I've actually remove the one helper function, since it is better expressed using this constructor.